### PR TITLE
Add support for a delayed activation flow for Thunderbolt

### PR DESCRIPTION
**Pull request by superm1**
This allows delaying the activation of Thunderbolt firmware until
shutdown/reboot or when the dock is unplugged.

This functionality requires features in the kernel:
https://lore.kernel.org/linux-usb/20200622143035.25327-1-mario.limonciello@dell.com/T/#t

Matrix of cases to support:

* Distro Old Linux kernel (doesn't support authenticate on disconnect)

  - WD19TB: Should have `skips-restart` flag set
    No flush or activate features called in `thunderbolt` plugin.
    `dell_dock` plugin will activate at end of composite update

  - All other devices: Shouldn't have flags set
    Should authenticate in Thunderbolt plugin.
    `1 > nvm_authenticate`

* Distro New Linux kernel (supports authenticate on disconnect)

  - WD19TB: Should have `usable-during-update` flag set but not `skips-restart`
    Should flush image to SPI in `thunderbolt` plugin
    `2 > nvm_authenticate_on_disconnect`
    Should configure TBT device for authenticate on disconnect
    `1 > nvm_authenticate_on_disconnect`
    `dell_dock` plugin will configure dock for authenticate on disconnect

  - All other devices: Shouldn't have flags set
    Should authenticate in `thunderbolt` plugin.
    `1 > nvm_authenticate`

* ChromeOS (supports authenticate on disconnect)

  - `thunerbolt.conf` will have `DelayedActivation=true`.

  - WD19TB: Should have `usable-during-update` flag set but not `skips-restart`
    Should flush image to SPI in `thunderbolt` plugin
    `2 > nvm_authenticate_on_disconnect`
    Should configure device for authenticate on disconnect
    `1 > nvm_authenticate_on_disconnect`
    `dell_dock` plugin will configure dock for authenticate on disconnect

  - All other devices: Should have both `usable-during-update` and `skips-restart` set
    Should flush image to SPI in `thunderbolt` plugin
    `2 > nvm_authenticate`
    Will activate upon logout/shutdown/reboot
    `1 > nvm_authenticate`

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
CC @djcampello

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 1]
```diff
@@ -178,6 +179,7 @@ fu_thunderbolt_device_to_string (FuDevice *device, guint idt, GString *str)
 	fu_common_string_append_kb (str, idt, "Safe Mode", self->safe_mode);
 	fu_common_string_append_kb (str, idt, "Native mode", self->is_native);
 	fu_common_string_append_ku (str, idt, "Generation", self->gen);
+	fu_common_string_append_kv (str, idt, "AuthAttribute", self->authenticate_attribute);
```
Comment (gicmo) [inline plugins/thunderbolt/fu-thunderbolt-device.c:182]: I stumbled over this when I first read it but I cannot think how to better name that is now horrendously long. I was thinking maybe `AuthMethod`/`AuthMechanism`? Maybe just good as is.

File: plugins/dell-dock/fu-plugin-dell-dock.c [hunk 2]
```diff
@@ -201,15 +217,9 @@ fu_plugin_composite_cleanup (FuPlugin *plugin,
...
-	/* if thunderbolt is in the transaction it needs to be activated separately */
-	for (guint i = 0; i < devices->len; i++) {
-		FuDevice *dev = g_ptr_array_index (devices, i);
-		if (g_strcmp0 (fu_device_get_plugin (dev), "thunderbolt") == 0 &&
-		    fu_device_has_flag (dev, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION)) {
-			if (!fu_device_activate (dev, error))
-				return FALSE;
-			break;
-		}
+	if (needs_activation && dev != NULL) {
```
Comment (gicmo) [inline plugins/dell-dock/fu-plugin-dell-dock.c:220]: I dont think `needs_activation` can be non-`NULL` but `dev` be `NULL` but doesn't hurt.

Comment (superm1) [reply plugins/dell-dock/fu-plugin-dell-dock.c:220]: `needs_activation` is boolean

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 1 above)
Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-device.c:182]: thanks fixed

File: plugins/dell-dock/fu-plugin-dell-dock.c [hunk 3]
```diff
@@ -185,11 +185,27 @@ fu_plugin_composite_cleanup (FuPlugin *plugin,
...
 	if (parent == NULL)
 		return TRUE;
 
+	/* if thunderbolt is in the transaction it needs to be activated separately */
+	for (guint i = 0; i < devices->len; i++) {
+		dev = g_ptr_array_index (devices, i);
+		if (g_strcmp0 (fu_device_get_plugin (dev), "thunderbolt") == 0 &&
+		    fu_device_has_flag (dev, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION)) {
+			if (fu_device_has_flag (dev, FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE)) {
+				fu_dell_dock_ec_tbt_passive (parent);
```
[the remark this answers is not in the document]
Comment (superm1) [reply plugins/dell-dock/fu-plugin-dell-dock.c:201]: added a comment

File: plugins/dell-dock/fu-plugin-dell-dock.c (hunk 2 above)
Comment (gicmo) [reply plugins/dell-dock/fu-plugin-dell-dock.c:222]: Ah yes, sorry, I mean, `needs_activation` can not be `TRUE` and  `dev` be `NULL` at the same time. But being defensive is probably a better idea anyway!

Comment (superm1) [reply plugins/dell-dock/fu-plugin-dell-dock.c:222]: Yeah it's mostly to help prevent if some other code changes this logic in the future and there is a situation that could occur. Thanks!

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 4]
```diff
@@ -321,36 +323,45 @@ fu_thunderbolt_device_setup (FuDevice *device, GError **error)
...
+		self->auth_method = "nvm_authenticate";
+	}
+
 	/* success */
 	return TRUE;
 }
 
 static gboolean
-fu_thunderbolt_device_trigger_update (FuDevice *device, GError **error)
+_write_simple_sysfs (gchar *path, const gchar* val, GError **error)
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-device.c:344]: could we split this out in FuUdevDevice?

File: plugins/thunderbolt/thunderbolt.conf [hunk 5]
```diff
@@ -4,3 +4,6 @@
 # It's important that all backports from this kernel have been
 # made if using an older kernel
 MinimumKernelVersion=4.13.0
+
+# Forces delaying activation until shutdown/logout/reboot
+DelayedActivation=false
```
Comment (hughsie) [inline plugins/thunderbolt/thunderbolt.conf:9]: why a config option?

Comment (superm1) [reply plugins/thunderbolt/thunderbolt.conf:9]: ChromeOS will want to turn this on for everything with a guarantee that they always run activation on logout/shutdown/reboot.  I didn't think this made sense for generically doing in all distros.  

If we want to shift fwupd to behave this way, we can drop the config option (and some logic).  See the matrix of situations to support in the commit message for background.

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 4 above)
Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-device.c:344]: You mean make a "write simple sysfs" method for FuUdevDevice that we call?

Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: I just pushed what I think you mean.

**Comment by superm1** [comment]
V2 of kernel patches needed for this: https://lore.kernel.org/linux-usb/20200622185758.28145-1-mario.limonciello@dell.com/T/#t

**Pull request merged**